### ISSUE TITLE
Question regarding GIT docu/hints

### ISSUE THREAD

**Issue by mkaesemann** [larshp/abapGit#1726]
@larshp:
I have a stable version of the parallel serialization running in my system and have parallelized the object existence check as well. This gives me a huge boost in processing time and the next target that would be worth optimization is the GIT decode and delta processing.
I have had a look at the coding in `ZCL_ABAPGIT_GIT_PACK` and from what I can tell you are prcessing the received PACK file directly, which, while possible, does not allow paralleization, because you need the length of the decoded file to determine the index of the next.
As far as I know, GIT also has an index file (IDX) per PACK file, which would allowdirect addressing with offsets. But while I am able to find a little a little documentation on how the file is organized and how to parse it, I am unable to find how to acquire it.
Can you share any documentation or links that you may have used to create the request in `UPLOAD_PACK` and that might give some insight into how to request the IDX file as well?
Or did you have to reverse-engineer everything as well.
Thought it might be worth asking before I go through the git sourcecode 😉

**Comment by larshp** [comment]
Hi,
I've collected some technical links at http://docs.abapgit.org/development/technical-links.html

Most important for the question above is https://github.com/git/git/blob/master/Documentation/technical/pack-format.txt

As far as I have been able to figure out, it is not possible to process the packfile in parallel, and the index file is only for file storage, not part of the packformat/file

**Comment by larshp** [comment]
I guess the best optimization would be implementing https://github.com/larshp/abapGit/issues/357

**Comment by mkaesemann** [comment]
That's one option I have been looking at, but I don't think I'll have the time anytime soon. 
I was also thinking of simply parsing through the packfile once and using the git and zlib headers to construct an index first and then do the actual decoding in parallel tasks. But besides a lack of time I also bumped into the obstacle that ABAP does not have bit-shift operations and not the best bit-processing operators, so handling the zlib headers might prove to be more time-consuming that the actual benefit of parallel decoding.
Hopefully my workload at work will decrease again and I'll have some time to take a closer look.
Haven't gotten around to giving you feedback on #1714 either 😭

**Comment by larshp** [comment]
I dont think the decoding can be done in parallel just looking at the zlib headers, a zlib binary body can contain the same hex sequence as the zlib header, and the length is not known until decoding is done.

I'll probably just move ahead with #1714, when time allows, I easily get distracted, like the refactoring of the git classes yesterday.

Can we close this issue for now?

**Comment by mkaesemann** [comment]
I was thinking more on the line of "parse through the pack once and detect the headers and end of files to create a lookup index prior to actually decoding".
In theory this would require O(1) for parsing through the pack once, while skipping data and looking for the end-of-block markers of zlib. Since we currently perform a decompression and a compression to parse the pack and compression is much slower in DEFLATE than decompression, the hope is that by skipping the decompression & compression and moving them to parallel tasks, that this makes up for the extra time needed to build the index.
Naturally this only works if I can find a way to keep the overhead for parsing through the stream low, which needs some more investigation upfront.

But that was not your question 😉 
I will close this issue for now and should I have enough time to look at this further and find that it is worth the effort, I will create a pull request for it anyway.
